sympy/solvers/solveset.py

```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+                            FiniteSet(value_res) - value_complement
+            result.append(res_copy)
+        result_finiteset = result
+
+    # convert to ordered tuple
+    result = S.EmptySet
+    for r in result_finiteset:
+        temp = [r[sym] for sym in all_symbols]
+        result += FiniteSet(tuple(temp))
+    return result
+
+
+def nonlinsolve(system, symbols):
+    r"""
+    Solve system of N non linear equations with M variables, which means both
+    under and overdetermined systems are supported. Positive dimensional
+    system is also supported (A system with infinitely many solutions is said
+    to be positive-dimensional). In Positive dimensional system solution will
+    be dependent on at least one symbol. If system doesn't have real solution
+    then complex solution will be returned in general form (in terms of `_n`,
+    where `_n` is any real number). The possible number of solutions is zero,
+    one or infinite.
+
+    Parameters
+    ==========
+
+    system : list of equations
+        The target system of equations
+    symbols : list of Symbols
+        symbols should be given as a sequence eg. list
+
+    Returns
+    =======
+
+    A FiniteSet of ordered tuple of values of `symbols` for which the `system`
+    has solution. Order of value is same as symbols order in the
+    `all_symbols`list.
+
+    Please note that general FiniteSet is unordered, the solution returned
+    here is not simply a FiniteSet of solutions, rather it is a FiniteSet of
+    ordered tuple, i.e. the first & only argument to FiniteSet is a tuple of
+    solutions, which is ordered, & hence the returned solution is ordered.
+
+    Also note that solution could also have been returned as an ordered tuple,
+    FiniteSet is just a wrapper `{}` around the tuple. It has no other
+    significance except for the fact it is just used to maintain a consistent
+    output format throughout the solveset.
+
+    For the given set of Equations, the respective input types
+    are given below:
+
+    .. math:: x*y - 1 = 0
+    .. math:: 4*x**2 + y**2 - 5 = 0
+
+    `system  = [x*y - 1, 4*x**2 + y**2 - 5]`
+    `symbols = [x, y]`
+
+    Examples
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y, z = symbols('x, y, z', real=True)
+    >>> nonlinsolve([x*y - 1, 4*x**2 + y**2 - 5], [x, y])
+    {(-1, -1), (-1/2, -2), (1/2, 2), (1, 1)}
+
+    * Positive dimensional system and complements:
+
+    >>> from sympy import pprint
+    >>> from sympy.polys.polytools import is_zero_dimensional
+    >>> a, b, c, d = symbols('a, b, c, d', real=True)
+    >>> foo =  a + b + c + d
+    >>> bar = a*b + b*c + c*d + d*a
+    >>> foo_bar = a*b*c + b*c*d + c*d*a + d*a*b
+    >>> bar_foo = a*b*c*d -1
+    >>> system = [foo, bar, foo_bar, bar_foo]
+    >>> is_zero_dimensional(system)
+    False
+    >>> pprint(nonlinsolve(system, [a, b, c, d]), use_unicode=False)
```

Comment:
The API seems to be counter-intuitive to me.
Don't you agree that all the values should have the same type?